text/0078-merkleized-metadata.md

```diff
@@ -0,0 +1,394 @@
+    len: u32,
+    type_param: TypeRef,
+}
+
+struct BitSequence {
+    num_bytes: u8,
+    least_significant_bit_first: bool,
+}
+
+struct EnumerationVariant {
+    name: String,
+    fields: Vec<Field>,
+    index: Compact<u32>,
+}
+
+enum TypeRef {
+    Bool,
+    Char,
+    Str,
+    U8,
+    U16,
+    U32,
+    U64,
+    U128,
+    U256,
+    I8,
+    I16,
+    I32,
+    I64,
+    I128,
+    I256,
+    CompactU8,
+    CompactU16,
+    CompactU32,
+    CompactU64,
+    CompactU128,
+    CompactU256,
+    Void,
+    PerId(Compact<u32>),
+}
+```
+
+The `Type` declares the structure of a type. The `type` has the following fields:
+
+- `path`: A `path` declares the position of a type locally to the place where it is defined. The `path` is not globally unique, this means that there can be multiple types with the same `path`.
+- `type_def`: The high-level type definition, e.g. the type is a composition of fields where each field has a type, the type is a composition of different types as `tuple` etc.
+- `type_id`: The unique identifier of this type.
+
+Every `Type` is composed of multiple different types. Each of these "sub types" can reference either a full `Type` again or reference one of the primitive types. This is where `TypeRef` becomes relevant as the type referencing information. To reference a `Type` in the type information, a unique identifier is used. As primitive types can be represented using a single byte, they are not put as separate types into the type information. Instead the primitive types are directly part of `TypeRef` to not require the overhead of referencing them in an extra `Type`. The special primitive type `Void` represents a type that encodes to nothing and can be decoded from nothing. As FRAME doesn't support `Compact` as primitive type it requires a more involved implementation to convert a FRAME type to a `Compact` primitive type. SCALE only supports `u8`, `u16`, `u32`, `u64` and `u128` as `Compact` which maps onto the primitive type declaration in the RFC. One special case is a `Compact` that wraps an empty `Tuple` which is expressed as primitive type `Void`.
+
+The `TypeDef` variants have the following meaning:
+
+- `Composite`: A `struct` like type that is composed of multiple different fields. Each `Field` can have its own type. A `Composite` with no fields is expressed as primitive type `Void`.
+- `Enumeration`: Stores a `EnumerationVariant`. A `EnumerationVariant` is a struct that is described by a name, an index and a vector of `Field`s, each of which can have it's own type. Typically `Enumeration`s have more than just one variant, and in those cases `Enumeration` will appear multiple times, each time with a different variant, in the type information. `Enumeration`s can become quite large, yet usually for decoding a type only one variant is required, therefore this design brings optimizations and helps reduce the size of the proof. An `Enumeration` with no variants is expressed as primitive type `Void`.
```

Comment:
I think it is worth mentioning, as below that each of the `Enumeration` instances will have the same `TypeRef` lookup id. This detail helped me understand how this worked.

> For `Enumeration`s all variants have the same unique identifier, while they are represented as multiple type information. All variants need to have the same unique identifier as the reference doesn't know which variant will appear in the actual encoded data.

Reply:
I mean this is explained in `Generating TypeRef`.